File: app/src/main/java/com/ishmah/musichub/db/CachedImageDao.java
package com.ishmah.musichub.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CachedImageDao {

    private final SQLiteDatabase db;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public CachedImageDao(Context context) {
        db = new DatabaseHelper(context).getWritableDatabase();
    }

    public void saveCache(String queryKey, String coverUrl,
                          String artistPhoto, Runnable onDone) {
        executor.execute(() -> {
            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_QUERY_KEY, queryKey);
            values.put(DatabaseHelper.COL_COVER_URL, coverUrl);
            values.put(DatabaseHelper.COL_ARTIST_PHOTO_URL, artistPhoto);
            values.put(DatabaseHelper.COL_CACHED_AT,
                    String.valueOf(System.currentTimeMillis()));
            db.insertWithOnConflict(DatabaseHelper.TABLE_CACHED_IMAGES,
                    null, values, SQLiteDatabase.CONFLICT_REPLACE);
            if (onDone != null) onDone.run();
        });
    }

    public Map<String, String> getCache(String queryKey) {
        Cursor cursor = db.query(DatabaseHelper.TABLE_CACHED_IMAGES,
                null,
                DatabaseHelper.COL_QUERY_KEY + " = ?",
                new String[]{queryKey},
                null, null, null);
        if (!cursor.moveToFirst()) {
            cursor.close();
            return null;
        }
        Map<String, String> cache = new HashMap<>();
        cache.put("query_key", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_QUERY_KEY)));
        cache.put("cover_url", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_COVER_URL)));
        cache.put("artist_photo", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_ARTIST_PHOTO_URL)));
        cache.put("cached_at", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CACHED_AT)));
        cursor.close();
        return cache;
    }

    public void clearOldCache(Runnable onDone) {
        executor.execute(() -> {
            long sevenDaysAgo = System.currentTimeMillis() - 7 * 24 * 60 * 60 * 1000L;
            db.delete(DatabaseHelper.TABLE_CACHED_IMAGES,
                    DatabaseHelper.COL_CACHED_AT + " < ?",
                    new String[]{String.valueOf(sevenDaysAgo)});
            if (onDone != null) onDone.run();
        });
    }

    public void clearAllCache(Runnable onDone) {
        executor.execute(() -> {
            db.delete(DatabaseHelper.TABLE_CACHED_IMAGES, null, null);
            if (onDone != null) onDone.run();
        });
    }
}

File: app/src/main/java/com/ishmah/musichub/db/CachedImageDao.kt
package com.ishmah.musichub.db

import android.content.ContentValues
import android.content.Context
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class CachedImageDao(context: Context) {

    private val db = DatabaseHelper(context).writableDatabase
    private val executor: Executor = Executors.newSingleThreadExecutor()

    // Simpan cache gambar
    fun saveCache(
        queryKey: String,
        coverUrl: String,
        artistPhoto: String,
        onDone: (() -> Unit)? = null
    ) {
        executor.execute {
            val values = ContentValues().apply {
                put(DatabaseHelper.COL_QUERY_KEY, queryKey)
                put(DatabaseHelper.COL_COVER_URL, coverUrl)
                put(DatabaseHelper.COL_ARTIST_PHOTO_URL, artistPhoto)
                put(DatabaseHelper.COL_CACHED_AT, System.currentTimeMillis().toString())
            }
            db.insertWithOnConflict(
                DatabaseHelper.TABLE_CACHED_IMAGES,
                null,
                values,
                android.database.sqlite.SQLiteDatabase.CONFLICT_REPLACE
            )
            onDone?.invoke()
        }
    }

    // Ambil cache berdasarkan query key
    fun getCache(queryKey: String): Map<String, String>? {
        val cursor = db.query(
            DatabaseHelper.TABLE_CACHED_IMAGES,
            null,
            "${DatabaseHelper.COL_QUERY_KEY} = ?",
            arrayOf(queryKey),
            null, null, null
        )
        if (!cursor.moveToFirst()) {
            cursor.close()
            return null
        }
        val cache = mapOf(
            "query_key" to cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_QUERY_KEY)),
            "cover_url" to cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_COVER_URL)),
            "artist_photo" to cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_ARTIST_PHOTO_URL)),
            "cached_at" to cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CACHED_AT))
        )
        cursor.close()
        return cache
    }

    // Hapus cache yang sudah lama (lebih dari 7 hari)
    fun clearOldCache(onDone: (() -> Unit)? = null) {
        executor.execute {
            val sevenDaysAgo = (System.currentTimeMillis() - 7 * 24 * 60 * 60 * 1000).toString()
            db.delete(
                DatabaseHelper.TABLE_CACHED_IMAGES,
                "${DatabaseHelper.COL_CACHED_AT} < ?",
                arrayOf(sevenDaysAgo)
            )
            onDone?.invoke()
        }
    }

    // Hapus semua cache
    fun clearAllCache(onDone: (() -> Unit)? = null) {
        executor.execute {
            db.delete(DatabaseHelper.TABLE_CACHED_IMAGES, null, null)
            onDone?.invoke()
        }
    }
}